src/seshat.erl

```diff
@@ -2,175 +2,380 @@
+                              Labels = case LabelFormat of
+                                           as_map ->
+                                               LabelsAsMap;
+                                           as_binary ->
+                                               LabelsAsBinary
+                                       end,
+                              format_fields(Fields, CRef, Labels, Acc);
+                          false ->
+                              %% skip filtered-out and unlabeled entries
+                              Acc
+                      end
               end, #{}, seshat_counters_server:get_table(Group)).
 
-%% internal
+format_fields(Fields, CRef, Labels, Acc) ->
+    lists:foldl(
+        fun ({Name0, Index, Type, Help}, Acc0) ->
+                PromType = prometheus_type(Type),
+                Name = with_prometheus_suffix(atom_to_binary(Name0), Type),
+                InitialMetric = #{type => PromType,
+                                    help => Help,
+                                    values => #{}},
+                MetricAcc = maps:get(Name, Acc0, InitialMetric),
+                ValuesAcc = maps:get(values, MetricAcc),
+                ComputedValue = case Type of
+                                    {_, ratio} ->
+                                        counters:get(CRef, Index) / 100;
+                                    {_, time_ms} ->
+                                        counters:get(CRef, Index) / 1000; % ms to s
+                                    _ ->
+                                        counters:get(CRef, Index) * 1.0 % ensure float
+                                end,
+                ValuesAcc1 = ValuesAcc#{Labels => ComputedValue},
+                MetricAcc1 = MetricAcc#{values => ValuesAcc1},
+                Acc0#{Name => MetricAcc1}
+        end, Acc, Fields).
 
-resolve_fields(Fields) when is_list(Fields) ->
+%% @doc Return the metadata for the fields
+%% When creating a set of metrics with seshat:new/3 or seshat:new/4,
+%% metadata about the metrics/counters has to be provided either
+%% as a list or as a reference to a persistent term with the list.
+%% The latter is recommended when a lot of similar metrics need to be stored
+%% (eg. many instances of the same component emit the same set of metrics)
+-spec resolve_fields_spec(fields_spec()) -> [field_spec()].
+resolve_fields_spec(Fields = FieldsSpec) when is_list(FieldsSpec) ->
     Fields;
-resolve_fields({persistent_term, PTerm}) ->
+resolve_fields_spec({persistent_term, PTerm}) ->
     %% TODO error handling
     persistent_term:get(PTerm).
 
-register_counter(Group, Name, Ref, Fields) ->
+-spec register_counter(group(), id(), counters:counters_ref(), fields_spec(), labels()) ->
+    ok.
+register_counter(Group, Id, CRef, FieldsSpec, Labels) when is_map(Labels) ->
     TRef = seshat_counters_server:get_table(Group),
-    true = ets:insert(TRef, {Name, Ref, Fields}),
+    Entry = #entry{
```

Comment:
no reason to use the space next to { for the id field and align the rest accordingly.